bigtable/client/internal/endian.h

```diff
@@ -0,0 +1,150 @@
+                               (((value)&0x0000000000ff0000ull) << 24) |
+                               (((value)&0x000000000000ff00ull) << 40) |
+                               (((value)&0x00000000000000ffull) << 56));
+
+  return return_value;
+}
+#endif
+#endif
+#endif
+
+inline bigtable::bigendian64_t byteswap64(bigtable::bigendian64_t value) {
+#ifdef _MSC_VER
+  return bigtable::bigendian64_t(_byteswap_uint64(value.get()));
+#elif defined(__GNUC__) || defined(__clang__)
+  return bigtable::bigendian64_t(__builtin_bswap64(value.get()));
+#else
+  return bigtable::bigendian64_t(explicit_swap64(value.get()));
+#endif
+}
+
+constexpr char c[sizeof(int)] = {1};
+constexpr bool IsBigEndian() {
+  return c[0] != 1;  // ignore different type comparison
+}
+
+template <typename T>
+struct encoder {
+  static std::string encode(T const& value);
+  static T decode(std::string const& value);
+};
+
+template <>
+struct encoder<bigtable::bigendian64_t> {
+  /**
+   * Convert BigEndian numeric value into a string of bytes and return it.
+   *
+   * Google Cloud Bigtable stores arbitrary blobs in each cell. These
+   * blobs are stored in a cell as a string of bytes. Values need to be
+   * converted to/from these cell blobs bytes to be used in application.
+   * This function is used to convert BigEndian 64 bit numeric value into
+   * string of bytes, so that it could be stored as a cell blob.
+   * For this conversion we consider the char size is 8 bits.
+   * So If machine architecture does not define char as 8 bit then we
+   * raise the assert.
+   *
+   * @param value reference to a number need to be converted into
+   *        string of bytes.
+   */
+  static std::string encode(bigtable::bigendian64_t const& value) {
+    static_assert(std::numeric_limits<unsigned char>::digits == 8,
+                  "This code assumes char is an 8-bit number");
+
+    std::string big_endian_string("", 8);
+    if (IsBigEndian()) {
+      std::memcpy((void*)big_endian_string.data(), &value, sizeof(value));
+    } else {
+      bigtable::bigendian64_t swapped_value = byteswap64(value);
+      std::memcpy((void*)big_endian_string.data(), &swapped_value,
+                  sizeof(swapped_value));
+    }
+    return big_endian_string;
+  }
+
+  /**
+   * Convert BigEndian string of bytes into BigEndian numeric value and
+   * return it.
+   *
+   * Google Cloud Bigtable stores arbitrary blobs in each cell. These
+   * blobs are stored in a cell as a string of bytes. Values need to be
+   * converted to/from these cell blobs bytes to be used in application.
+   * This function is used to convert string of bytes into BigEndian 64
+   * bit numeric value. so that it could be used in application.
+   * For this conversion we consider the char size is 8 bits.
+   * So If machine architecture does not define char as 8 bit then we
+   * raise the std::range_error.
+   *
+   * @param value reference to a string of bytes need to be converted into
```

Comment:
I can see that is a reference to a string of bytes, this is a case where I would remove the `@param`.